Allow Tooling API clients to invoke `--help` and `--version`

usv240 opened the pull request:
<!--- The issue this PR addresses -->
Fixes #14116

### Context
This PR implements support for `--help`, `--version`, and `--show-version` command-line arguments in the Tooling API.

Previously, providing these arguments via `longRunningOperation.withArguments(...)` would often fail (e.g., "Task not found" errors) or be ignored because the Tooling API treated the request as a standard build execution rather than a meta-information request.

**Implementation Details:**
- Introduced `VersionHelpConsumerActionExecutor` to the connection chain.
- The executor intercepts execution if `--help`, `--version`, or `--show-version` are detected.
- It explicitly clears tasks and arguments to prevent build failures.
- It queries the `Help` model (for `--help`) or `BuildEnvironment` model (for `--version`) to generate the correct output.
- For `--show-version`, it prints the version information and then proceeds with the original build action.

### Contributor Checklist
- [x] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md).
- [x] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [x] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [x] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team.
- [x] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective.
- [x] Provide unit tests (under `<subproject>/src/test`) to verify logic.
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes.
- [x] Ensure that tests pass sanity check: `./gradlew sanityCheck`.
- [x] Ensure that tests pass locally: `./gradlew :tooling-api:quickTest`.

donat commented on platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r940/VersionHelpConsumerCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,84 @@
+@TargetGradleVersion(">=9.4.0")
+class VersionHelpConsumerCrossVersionSpec extends ToolingApiSpecification {
+
+    def "prints version info and ignores tasks when --version is present"() {
```
❌ Missing test coverage for old target Gradle versions.

donat commented on platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/VersionHelpConsumerActionExecutor.java:
```diff
@@ -0,0 +1,256 @@
+import java.util.Collections;
+import java.util.List;
+
+public class VersionHelpConsumerActionExecutor implements ConsumerActionExecutor {
```
❌ Although this kind of works, I find it quite hack-ish. I've spent quite some time and looked into the details. The main problem is that this ActionExecutor intercepts the TAPI requests at the wrong abstraction level: outside of where the project connection is available.

The main issue is that here we don't know anything about the provider side (ie the target Gradle distribution). Currently it seems sufficient that we simply catch the exception when an unsupported Gradle version is being used. At the same time, the plan is to backport the `--help` and the `--version` rendering for older Gradle versions that is impossible here.

I propose the following: move the interception logic into `org.gradle.tooling.internal.consumer.connection.AbstractConsumerConnection`. 
- Maybe extract the interception logic to a subclass between `AbstractConsumerConnection` and it's existing immediate subclass.
- There, you have access to `VersionDetails` which you can use to separate the supported/unsupported cases.
- You can do the interceptions only for task execution.
- For model loading, launching tests you could look into how to improve existing functionality
  - Don't fail model loading/test execution when `--help` or `--version` is listed as arguments
  - Instead emit a warning that the argument list was cleaned up.

Later (in a separate PR) we can add static content for older Gradle versions in the new ConsumerConnection class.

donat commented on testing/internal-testing/src/main/resources/valid-precondition-combinations.csv:
```diff
@@ -182,3 +182,4 @@ IntegTestPreconditions$UnsupportedWorkerJavaHomeAvailable
 IntegTestPreconditions$Java21HomeAvailable,IntegTestPreconditions$Java17HomeAvailable,IntegTestPreconditions$NotEmbeddedExecutor
 UnitTestPreconditions$NotAlpine
 UnitTestPreconditions$NotWindows,UnitTestPreconditions$NotAlpine
+SmokeTestPreconditions$Jdk17Available,UnitTestPreconditions$Jdk9OrLater,IntegTestPreconditions$NotConfigCached
```
🤔 Seems unrelated

donat commented on testing/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleBuildToolingApiSmokeTest.groovy:
```diff
@@ -68,4 +68,62 @@ class GradleBuildToolingApiSmokeTest extends AbstractGradleceptionSmokeTest {
         ].combinations()
     }
+
+    def "can run with --version"() {
```
❌ We probably don't need this. The cross-version coverage is quite enough.

donat commented on testing/smoke-test/src/testFixtures/groovy/org/gradle/test/preconditions/SmokeTestPreconditions.groovy:
```diff
@@ -16,11 +16,22 @@
 
 class SmokeTestPreconditions {
 
+    static class Jdk17Available implements TestPrecondition {
```
Same as above.

donat reviewed (changes requested):
I'm confused now. The last commit on this PR removed the behavior that we expect:

If the user runs the build with `--help` or `--version` the tasks should be ignored and the corresponding help or version info should be printed on the console.

Please restore that behavior.

donat commented on platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r940/VersionConsumerCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        then:
+        result.output.contains("Gradle")
```
❌ Please add a bit more specific assertion for the output. The `Gradle` label might appear in the build output outside of the version information.

donat commented on platforms/ide/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r940/VersionConsumerCrossVersionSpec.groovy:
```diff
@@ -0,0 +1,52 @@
+
+@ToolingApiVersion(">=9.4.0")
+class VersionConsumerCrossVersionSpec extends ToolingApiSpecification {
+
```
❌ missing test coverage:

What happens for older target Gradle versions
What happens when --help is passed to the TestLauncher/BuildAction/ModelBuilder (ie to all entry points)?

donat commented on platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/AbstractConsumerConnection.java:
```diff
@@ -61,24 +72,145 @@ public ConnectionVersion4 getDelegate() {
+                    operationParameters.getStandardOutput().write(output.getBytes(StandardCharsets.UTF_8));
+                } catch (UnsupportedMethodException e) {
+                    // Fallback for older Gradle versions that don't have getVersionInfo()
+                    String fallback = "Gradle " + env.getGradle().getGradleVersion() + "\n";
```
❌ I think version should behave the same as help for older versions. We should print:

`Failed to obtain version information for the selected Gradle version`.

donat commented on platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/AbstractConsumerConnection.java:
```diff
@@ -61,24 +72,145 @@ public ConnectionVersion4 getDelegate() {
+            "--show-version".equals(arg) || "-V".equals(arg));
+
+        if (removed && warn) {
+            LOGGER.warn("The Tooling API does not support --help, --version or --show-version arguments for this operation. These arguments have been ignored.");
```
Note in the tests you can assert the log messages via the `result.error` property and via the associated `assertHasErrorOutput()` helper method.

donat commented on platforms/ide/tooling-api/src/main/java/org/gradle/tooling/internal/consumer/connection/TestExecutionConsumerConnection.java:
```diff
@@ -34,7 +34,7 @@ public TestExecutionConsumerConnection(ConnectionVersion4 delegate, ModelMapping
 
     @Override
-    public void runTests(final TestExecutionRequest testExecutionRequest, ConsumerOperationParameters operationParameters) {
+    protected void doRunTests(final TestExecutionRequest testExecutionRequest, ConsumerOperationParameters operationParameters) {
```
❌ unnecessary change?

usv240 commented:
Hi @donat,

I have a comprehensive implementation ready locally that addresses all your review items including:

- Created separate HelpAndVersionHandlingConsumerConnection class
- Broke up handleHelpOrVersion into smaller methods
- Reverted the doRun/doRunTests renames
- Added test coverage for older Gradle versions and all entry points

I see you pushed some fixes - would you prefer I force push my complete implementation, or should we coordinate differently?"

donat commented:
@usv240 Thank you for the contribution!

usv240 commented:
Hi @donat,

Thank you so much for your patience and guidance throughout this PR! I really appreciate you taking the time to explain the architecture decisions.

I learned a lot about the Tooling API internals, cross-version testing, and how to structure code for maintainability in a large codebase like Gradle.

Sorry for the delay in my responses towards the end. I'll be more responsive going forward. 
Looking forward to contributing more to Gradle!
